File: StackWithMin/StackWithMin.cpp
#include <iostream>
#include <stack>

using std::cout;
using std::endl;
using std::stack;

class StackWithMin {
  public:
    StackWithMin();
    void push(int item);
    int pop();
    int getMin();
    int size();
    bool empty();
    int top();
  private:
    stack<int> minStack;
    stack<int> actualStack;
};

StackWithMin::StackWithMin() {}

void StackWithMin::push(int item) {
  actualStack.push(item);
  if (minStack.empty() || item <= minStack.top()) {
    minStack.push(item);
  }
}

int StackWithMin::pop() {
  int item = actualStack.top();
  actualStack.pop();
  if (item == minStack.top()) {
    minStack.pop();
  }
  return item;
}

int StackWithMin::getMin() {
  return minStack.top();
}

int StackWithMin::size() {
  return actualStack.size();
}

bool StackWithMin::empty() {
  return (actualStack.size() == 0);
}

int StackWithMin::top() {
  return actualStack.top();
}

int main() {
  StackWithMin first;
  for (int i = 0; i < 5; i++) {
    first.push(i);
  }
  if (!first.empty()) {
    cout << "Not empty" << endl;
  }
  first.pop();
  cout << "Top is " << first.top() << endl;
  cout << "Stack contains " << first.size() << " items" << endl;
  cout << "The smallest element is " << first.getMin() << endl;
}

File: StackWithMin/StackWithMin.py
#!/usr/bin/python

class StackWithMin:
  def __init__(self):
    self.actualStack = []
    self.minStack = []

  def push(self, item):
    self.actualStack.append(item)
    if not self.minStack or item <= self.minStack[-1]:
      self.minStack.append(item)

  def pop(self):
    item = self.actualStack.pop()
    if item == self.minStack[-1]:
      self.minStack.pop()
    return item

  def getMin(self):
    return self.minStack[-1]

  def __len__(self):
    return len(self.actualStack)

  def empty(self):
    return not self.actualStack

  def top(self):
    return self.actualStack[-1]

first = StackWithMin()
for i in xrange(5):
  first.push(i)
if not first.empty():
  print "Not empty"
first.pop()
print "Top is", first.top()
print "Stack contains", len(first), "items"
print "The smallest element is", first.getMin()
